ha/fault_tolerance/rest_api.py

```diff
@@ -0,0 +1,255 @@
+
+            if isinstance(resp, web.StreamResponse):
+                return resp
+
+            status = 200
+            if isinstance(resp, Response):
+                status = resp.rc()
+                resp_obj = {'response_body': resp.output(), 'status_code': status}
+                Log.info(f"Response = {resp_obj}")
+
+                if not 200 <= status <= 299:
+                    Log.error(f"Error: ({status}):{resp_obj['response_body']}")
+            else:
+                resp_obj = resp
+                Log.info(f"Response = {resp_obj}")
+
+            return CcRestApi.json_response(resp_obj, status)
+
+        # TODO: CORTX-30931 Handle multiple exceptions, below is the just 1 added
+        except (ConcurrentCancelledError, AsyncioCancelledError) as e:
+            Log.warn(f"Client cancelled call for {request.method} {request.path}")
+            # TODO: CORTX-30931 use CcRequestCancelled return proper response
+            return CcRestApi.json_response(CcRestApi.error_response(Exception("Call cancelled by client"),
+                                            request = request, request_id = request_id), status=499)
+
+    @staticmethod
+    async def _stop():
+        """
+        This function stops the REST Server.
+        cancels all the running tasks, stops site and event loop
+        """
+        CcRestApi.__is_shutting_down = True
+        for task in asyncio.Task.all_tasks():
+            if task != asyncio.Task.current_task():
+                task.cancel()
+        await CcRestApi._site.stop()
+        CcRestApi._loop.stop()
+
+
+    @staticmethod
+    async def handle_signal(signame: str):
+        """
+        Signal handler callback function.
+        if SIGTERM, SIGINT is received it stops the REST server.
+        other all are logged and ignored.
+        Args:
+            signame (str): Singal name which is getting handled.
+        """
+        Log.info(f'Received signal: {signame}: {getattr(signal, signame)}')
+        if signame in ('SIGTERM', 'SIGINT'):
+            await CcRestApi._stop()
+        else:
+            Log.info(f'Signal: {signame}: {getattr(signal, signame)} is ignored.')
+
+    @staticmethod
+    async def _on_startup(app):
+        """
+        This function is executes on startup of CC Rest API Server.
+        Args:
+            app (web.Application): web Application object which is starting.
+        """
+        Log.info(f'REST API server {CcRestApi._site.name} startup')
+        # Note: Additional calls needs to be added to execute on startup of CC Rest API Server
+
+    @staticmethod
+    async def _on_shutdown(app):
+        """
+        This function is executes on shutdown of CC Rest API Server.
+        Args:
+            app (web.Application): web Application object which is shuting down.
+        """
+        Log.info(f'REST API server {CcRestApi._site.name} shutdown')
+        # Note: Additional calls needs to be added to execute on shutdown of CC Rest API Server
```

Comment:
same as above

Reply:
In current implementation of REST API we don't have any such task which needs to be run in background while application shutdown that's why we kept it as note. In future if any such call/task required while shutdown of application we will add those call/task here.